### ISSUE TITLE
Proper support for functional discriminators

### ISSUE THREAD

**Issue by samuelcolvin** [pydantic/pydantic#7462]
### Initial Checks

- [X] I have searched Google & GitHub for similar requests and couldn't find anything
- [X] I have read and followed [the docs](https://docs.pydantic.dev) and still think this feature is missing

### Description

this is supported in pydantic-core, see [here](https://github.com/pydantic/pydantic-core/blob/f6b14cce22b1539727ed0e0a47d2c680c79c1692/tests/validators/test_tagged_union.py#L430-L447) but doesn't seem to be well supported in Pydantic V2.

I propose we add a new type:

```py
@dataclass
class Discriminator:
    discriminator: Callable[[Any], str] | str
    _: KW_ONLY
    # defaults to `{choice.__name__: choice for choice in choices}`
    choice_lookup: Callable[tuple[type[Any], ...], dict[str, type[Any]]] | None = None
```

Obviously `Discriminator` should behave the same as `Field(discriminator='...')` when it's used with a string.

Which could be used like this:

```py
from dataclasses import dataclass
from typing import Annotated

from pydantic import TypeAdapter, Discriminator


@dataclass
class City:
    name: str


@dataclass
class Country:
    name: str


def discriminator_function(obj: Any) -> str:
    if obj.name == 'London':
        return 'City'
    else:
        return 'Country'


ts = TypeAdapter(Annotated[City | Country, Discriminator(discriminator_function)])
```

Related to #7391 but deserved a better explanation.

### Affected Components

- [ ] [Compatibility between releases](https://docs.pydantic.dev/changelog/)
- [X] [Data validation/parsing](https://docs.pydantic.dev/usage/models/#basic-model-usage)
- [ ] [Data serialization](https://docs.pydantic.dev/usage/exporting_models/) - `.model_dump()` and `.model_dump_json()`
- [ ] [JSON Schema](https://docs.pydantic.dev/usage/schema/)
- [ ] [Dataclasses](https://docs.pydantic.dev/usage/dataclasses/)
- [ ] [Model Config](https://docs.pydantic.dev/usage/model_config/)
- [ ] [Field Types](https://docs.pydantic.dev/usage/types/) - adding or changing a particular data type
- [ ] [Function validation decorator](https://docs.pydantic.dev/usage/validation_decorator/)
- [ ] [Generic Models](https://docs.pydantic.dev/usage/models/#generic-models)
- [ ] [Other Model behaviour](https://docs.pydantic.dev/usage/models/) - `model_construct()`, pickling, private attributes, ORM mode
- [ ] [Plugins](https://docs.pydantic.dev/) and integration with other tools - mypy, FastAPI, python-devtools, Hypothesis, VS Code, PyCharm, etc.

**Comment by karlicoss** [comment]
Ah sorry if I was a bit unclear in https://github.com/pydantic/pydantic/issues/7391
Basically I'd like a way for any union to work (including any recursive union types, i.e. if it's a union field in a dataclass etc) without explicitly adding discriminators for them (I'm not sure if it somehow goes against pydantic philosophy though)

In this particular case I can see how your suggestion would work for serialization -- however not sure it's gonna work for deserializing (`ts.validate_python` call)? Perhaps you could add a suggested output of `ts.dump_python` when using a functional discriminator -- I think it might be a bit more clear then?

**Comment by sydney-runkle** [comment]
Side note -- looks like David has already made significant progress in this department: #6915.

Will touch base with the team on this tmrw morning.

**Comment by JacobHayes** [comment]
Would this (or something else) support dynamically registering members _that are not know when defining the `Union`_ (or equivalent)? Of course, all members would still have to be known by the time of (de)serialization (or perhaps dynamically registered by `discriminator_function` if possible).

---

For example, I've worked on 3 unrelated projects where a base class defines a couple fields but is then "dynamically" subclassed with more fields. In two of the projects, external users are creating the subclasses, hence can't be know ahead of time (but we can implement a plugin/registration mechanism to ensure they _are_ know by the time we try (de)serialization). The other project uses separate modules to avoid circular imports, so maintaining the `Union` definition each time we add a module/subclass is extra challenging.

The reason we'd like a discriminated union here, instead of just hinting with the base class, is because of these additional fields defined on the subclasses. If a wrapping model is just hinted with the base class, the extra fields are discarded.

**Comment by sydney-runkle** [comment]
Ta da! Closed by #7983 🥳
